src/ps/handlers/joins.ts

```diff
@@ -1,18 +1,62 @@
-import { PSAltCache, PSGames, PSSeenCache } from '@/cache';
+import { PSAltCache, PSGames, PSJoinphraseCache, PSSeenCache } from '@/cache';
 import { rename } from '@/database/alts';
 import { seeUser } from '@/database/seens';
 import { ChatError } from '@/utils/chatError';
 import { debounce } from '@/utils/debounce';
 import { fromHumanTime } from '@/utils/humanTime';
 import { toId } from '@/utils/toId';
 
+import type { PSMessage } from '@/types/ps';
 import type { Client } from 'ps-client';
 
+const minimumMessages: number = 5,
+	minimumTime: number = 30; // seconds
+
+interface jpState {
+	messageCount: number; // messages since last jp
+	lastTime: number; // timestamp of last jp
+}
+const jpStateMap: Partial<Record<string, jpState>> = {};
+
+export function otherHandler(message: PSMessage) {
+	if (message.isIntro) return;
+	if (!message.author || !message.author.userid || !message.target || message.author.id === message.parent.status.userid) return;
+	if (message.content.startsWith('|')) return;
+	const roomId = message.target.id;
+
+	for (const key in jpStateMap) {
+		if (!jpStateMap[key]) continue;
+
+		const rId = key.split('-')[1];
+		if (rId !== roomId) continue;
+
+		jpStateMap[key]!.messageCount++;
+	} // increment message count for each joinphrase in the room
```

Comment:
Performance wise this is a bit concerning; if there's a couple hundred JPs total, you're looping over every single one of them on every message

Instead, the state map should be grouped by rooms instead so you don't iterate over JPs that aren't for the room